core/capabilities/syncer.go

```diff
@@ -67,15 +86,412 @@ func NewRegistrySyncer(peerWrapper p2ptypes.PeerWrapper, registry core.Capabilit
+	// - remote capability DONs the current node is a part of.
+	// These need server-side shims.
+	myID := s.peerWrapper.GetPeer().ID()
+	myWorkflowDONs := []kcr.CapabilityRegistryDONInfo{}
+	remoteWorkflowDONs := []kcr.CapabilityRegistryDONInfo{}
+	myCapabilityDONs := []kcr.CapabilityRegistryDONInfo{}
+	remoteCapabilityDONs := []kcr.CapabilityRegistryDONInfo{}
+	for _, d := range publicDONs {
+		isDONForMyID := false
+		for _, peerID := range d.NodeP2PIds {
+			if peerID == myID {
+				isDONForMyID = true
+			}
+		}
+
+		if d.AcceptsWorkflows {
+			if isDONForMyID {
+				myWorkflowDONs = append(myWorkflowDONs, d)
+			} else {
+				remoteWorkflowDONs = append(remoteWorkflowDONs, d)
+			}
+		}
+
+		if len(d.CapabilityConfigurations) > 0 {
+			if isDONForMyID {
+				myCapabilityDONs = append(myCapabilityDONs, d)
+			} else {
+				remoteCapabilityDONs = append(remoteCapabilityDONs, d)
+			}
+		}
+	}
+
+	// Now, if my node is a workflow DON, let's setup any shims
+	// to external capabilities.
+	if len(myWorkflowDONs) > 0 {
+		for _, rcd := range remoteCapabilityDONs {
+			err := s.addRemoteCapabilities(ctx, rcd, state)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Finally, if I'm a capability DON, let's enable external access
+	// to the capability.
+	if len(myCapabilityDONs) > 0 {
+		for _, mcd := range myCapabilityDONs {
+			err := s.enableExternalAccess(ctx, myID, mcd, state, remoteWorkflowDONs)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func signersFor(don kcr.CapabilityRegistryDONInfo, state state) ([][]byte, error) {
+	s := [][]byte{}
+	for _, nodeID := range don.NodeP2PIds {
+		node, ok := state.IDsToNodes[nodeID]
+		if !ok {
+			return nil, fmt.Errorf("could not find node for id %s", nodeID)
+		}
+
+		s = append(s, node.Signer[:])
+	}
+
+	return s, nil
+}
+
+func toDONInfo(don kcr.CapabilityRegistryDONInfo) *capabilities.DON {
+	peerIDs := []p2ptypes.PeerID{}
+	for _, p := range don.NodeP2PIds {
+		peerIDs = append(peerIDs, p)
+	}
+	return &capabilities.DON{
+		ID:      fmt.Sprint(don.Id),
+		Members: peerIDs,
+		// TODO: risk of overflow, as don.F = uint32 and this F = uint8
+		// Remove the coercion when addressed in the contract.
+		F: uint8(don.F),
+	}
+}
+
+func (s *registrySyncer) addRemoteCapabilities(ctx context.Context, don kcr.CapabilityRegistryDONInfo, state state) error {
+	for _, c := range don.CapabilityConfigurations {
+		capability, ok := state.IDsToCapabilities[c.CapabilityId]
+		if !ok {
+			return fmt.Errorf("could not find capability matching id %s", c.CapabilityId)
+		}
+
+		switch capability.CapabilityType {
+		// trigger
+		case 0:
+			newTriggerFn := func(info capabilities.CapabilityInfo) (capabilityService, error) {
+				codec := streams.NewCodec(s.lggr)
+
+				signers, err := signersFor(don, state)
+				if err != nil {
+					return nil, err
+				}
+
+				aggregator := triggers.NewMercuryRemoteAggregator(
```

Comment:
Can this be based on capability ID? If not a "streams-trigger" then use a default MODE aggregator (not implemented yet...)